internal/sourcemap/metadata.go

```diff
@@ -0,0 +1,254 @@
+// under the License.
+
+package sourcemap
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/go-sourcemap/sourcemap"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+type MetadataCachingFetcher struct {
+	esClient         *elasticsearch.Client
+	set              map[Identifier]string
+	mu               sync.RWMutex
+	backend          Fetcher
+	logger           *logp.Logger
+	once             sync.Once
+	index            string
+	invalidationChan chan<- Identifier
+}
+
+func NewMetadataCachingFetcher(
+	c *elasticsearch.Client,
+	backend Fetcher,
+	index string,
+	invalidationChan chan<- Identifier,
+) *MetadataCachingFetcher {
+	return &MetadataCachingFetcher{
+		esClient:         c,
+		index:            index,
+		set:              make(map[Identifier]string),
+		backend:          backend,
+		logger:           logp.NewLogger(logs.Sourcemap),
+		invalidationChan: invalidationChan,
+	}
+}
+
+func (s *MetadataCachingFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
+	if len(s.set) == 0 {
+		// If cache is empty and we're trying to fetch a sourcemap
+		// make sure the initial cache population has ended.
+		s.once.Do(func() {
+			if err := s.sync(ctx); err != nil {
+				s.logger.Error("failed to fetch sourcemaps metadata: %v", err)
+			}
+		})
+	}
+
+	key := Identifier{
+		name:    name,
+		version: version,
+		path:    path,
+	}
+
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if _, found := s.set[key]; found {
+		// Only fetch from ES if the sourcemap id exists
+		return s.backend.Fetch(ctx, name, version, path)
+	}
+
+	// Try again
+	key.path = maybeParseURLPath(path)
+
+	if _, found := s.set[key]; found {
+		// Only fetch from ES if the sourcemap id exists
+		return s.backend.Fetch(ctx, name, version, path)
+	}
+
+	return nil, nil
+}
+
+// maybeParseURLPath attempts to parse s as a URL, returning its path component
+// if successful. If s cannot be parsed as a URL, s is returned.
+func maybeParseURLPath(s string) string {
+	url, err := url.Parse(s)
+	if err != nil {
+		return s
+	}
+	return url.Path
+}
+
+func (s *MetadataCachingFetcher) update(ms []Metadata) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, m := range ms {
+		if contentHash, ok := s.set[m.id]; ok {
+			// seen
+			delete(s.set, m.id)
+
+			// content hash changed, invalidate the sourcemap cache
+			if contentHash != m.contentHash {
+				s.invalidationChan <- m.id
+			}
+		}
+	}
+
+	// Loop for any unseed sourcemap
+	for k := range s.set {
+		// the sourcemap no longer exists in ES.
+		// invalidate the sourcemap cache.
+		s.invalidationChan <- k
+
+		// remove from metadata cache
+		delete(s.set, k)
+	}
+
+	for _, m := range ms {
+		s.set[m.id] = m.contentHash
+	}
```

Comment:
This logic seems a bit inefficient. 

Two ideas: 
(1) In the `sync` method, make `ms` a `map` with the key being `id`. Then iterate over `ms` once, check if the `id` is in the `ms` map, and act accordingly. Remove the items from the map an add the remaining map items afterwards. You save an iteration and a lot of writes to the cache. 

(2) You could use a cache with a time based eviction, that is set to something between `metadataFetcherInterval < cacheTimeout < 2*metadataFetcherInterval`. On eviction, the sourcemap key is added to the `s.invalidationChan`. On every update, the time is updated for the valid sourcemap keys.

Reply:
I believe option 1 is simpler and good enough.

Reply:
Thanks for spotting this! I've pushed some changes :bow: